typings/objection/index.d.ts

```diff
@@ -95,12 +96,38 @@ declare namespace Objection {
     skipValidation?: boolean;
   }
 
+  export type ValidatorArgs = {
+    ctx: object,
```

Comment:
Maybe a `Pojo` is a better type since it is ugly to access anything through an `object` type. The point of `ctx` is that implementors can store any data there.

Reply:
I just spent several hour investigating this, including having to isolate a [bug that crashed TSC](https://github.com/Microsoft/TypeScript/issues/20886), and I've made an interesting discovery.

I think you're right that the `object` type is useless. However, if you instead do this:

```ts
  export interface ValidatorContext { }

  export type ValidatorArgs = {
    ctx: ValidatorContext,
    model: Model,
    json: Pojo,
    options: ModelOptions
  }

  export class Validator {
    beforeValidate(args: ValidatorArgs): void;
    validate(args: ValidatorArgs): Pojo;
    afterValidate(args: ValidatorArgs): void;
  }
```

**AND** if you remove the `declare namespace Objection` and `export = Objection` from the typings, then you can do this in your app to get static type checking on the context:

```ts
declare module 'objection' {
    export interface ValidatorContext {
        foo: number;
    }
}
```

Why is Objection declared as a namespace anyway? I wonder if that's what's making it impossible to open Objection types for declaration merging. @mceachen?

Removing the outer namespace would allow apps to open QueryContext for extension to get static type checking.

Reply:
Woo! Another interesting discovery. We can have our cake and eat it too. If I declare:

```ts
export interface ValidatorContext {
    [key: string]: any;
}
```

Then I can assign `ctx.foo` to whatever I want, no problem.

But if the app opens that same Pojo ValidatorContext class as follows:

```ts
declare module 'objection' {
    export interface ValidatorContext {
        foo: number;
    }
}
```

Now suddenly `foo` has to be a number, but all other values remain unconstrained.

This would allow apps to decide whether to enforce static type checking, though it does require that the dev think ahead enough to add the decls necessary to get the checking.

So it appears that the only real problem is opening up the Objection namespace.

Reply:
The above `declare module 'objection'` decl does indeed open up the namespace! I thought I had tried that. Okay, so I've updated this PR with the following:

```ts
  export interface ValidatorContext extends Pojo { }

  export type ValidatorArgs = {
    ctx: ValidatorContext,
    model: Model,
    json: Pojo,
    options: ModelOptions
  }
```

I'd personally prefer to **have to** extend ValidatorContext in an app to be able to add app-specific properties, but in keeping with the rest of the typings, I went ahead and extended Pojo. That way at least it's possible for an app to restrict some properties.

So there's no need to remove the Objection namespace wrapper after all.

Reply:
Sorry for the delay. 

In reference to `type` vs `interface`: In general, `interface`s should be what you use for rather than a `type` alias. The [type alias docs](https://www.typescriptlang.org/docs/handbook/advanced-types.html#type-aliases) are fairly clear. (Note that the TypeScript 2.3 series would actually crash when compiling older versions of objection, because I hadn't internalized the fact that types were uncached by the compiler, and caused memory issues. The solution was to convert most type aliases to interfaces).

I'm happy to change any type alias in the typings to an interface.

Reply:
I should have thought to do that. I find myself frequently turning types into interfaces.

How do you feel about ValidatorContext extending Pojo? I think that's implementation inheritance -- ValidatorContext isn't always and forever a thing of type Pojo. I should nix that and put `[key: string]: any` back in, as you've done elsewhere.